enterprise/views.py

```diff
@@ -1137,3 +1138,136 @@ def post(self, request, enterprise_uuid, program_uuid):
+        PROGRAM_ENROLLMENT_VIEW_URL: ProgramEnrollmentView,
+        HANDLE_CONSENT_ENROLLMENT_VIEW_URL: HandleConsentEnrollment,
+    }
+
+    @staticmethod
+    def get_path_variables(**kwargs):
+        """
+        Get the base variables for any view to route to.
+
+        Currently gets:
+        - `enterprise_uuid` - the UUID of the enterprise customer.
+        - `course_id` - the ID of the course, if applicable.
+        - `program_uuid` - the UUID of the program, if applicable.
+        """
+        enterprise_customer_uuid = kwargs.get('enterprise_uuid', '')
+        course_id = kwargs.get('course_id', '')
+        program_uuid = kwargs.get('program_uuid', '')
+        return enterprise_customer_uuid, course_id or program_uuid
+
+    def eligible_for_direct_audit_enrollment(self, request, enterprise_customer, course_id):
+        """
+        Return whether a request is eligible for direct audit enrollment for a particular enterprise customer.
+
+        We check for the following criteria:
+        - The `audit` query parameter.
+        - The user's being routed to the course enrollment landing page.
+        - The customer's catalog contains the course in question.
+        """
+        request_has_audit_query_param = request.GET.get('audit')
+        is_course_enrollment_url = request.path == self.COURSE_ENROLLMENT_VIEW_URL.format(
+            enterprise_customer.uuid,
+            course_id
+        )
+        customer_catalog_contains_course = enterprise_customer.catalog_contains_course(course_id)
+        return request_has_audit_query_param and is_course_enrollment_url and customer_catalog_contains_course
+
+    def enroll_learner(self, request, enterprise_customer, course_id, mode):
+        """
+        Enroll a user into a course track, and register enterprise course enrollment & data sharing consent objects.
+        """
+        with transaction.atomic():
+            enterprise_customer_user, __ = EnterpriseCustomerUser.objects.get_or_create(
+                enterprise_customer=enterprise_customer,
+                user_id=request.user.id
+            )
+
+        enrollment_api_client = EnrollmentApiClient()
+
+        # Check to see if the user's already enrolled and we have an enterprise course enrollment to track it.
+        enrolled_in_course = enrollment_api_client.is_enrolled(request.user.username, course_id)
+        enterprise_course_enrollment_exists = EnterpriseCourseEnrollment.objects.filter(
+            enterprise_customer_user__enterprise_customer=enterprise_customer,
+            enterprise_customer_user__user_id=request.user.id,
+            course_id=course_id
+        ).exists()
+        if not (enrolled_in_course and enterprise_course_enrollment_exists):
+            # Directly enroll into the audit track.
+            EnterpriseCourseEnrollment.objects.update_or_create(
+                enterprise_customer_user=enterprise_customer_user,
+                course_id=course_id,
+            )
```

Comment:
Remove this `update_or_create` call.

Reply:
:+1: but I will be moving the new `get_or_create` all the way to the bottom, to ensure we safely enroll the user in the course itself first before creating the enterprise course enrollment object.